From 2ff526a4f206604e8ec99d93648f79c163e8a966 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:42:47 -0400
Subject: [infiniband] cxgb4: Fix incorrect BUG_ON conditions

Message-id: <1410475447-94392-29-git-send-email-svemuri@redhat.com>
Patchwork-id: 93883
O-Subject: [RHEL7.1 PATCH BZ 1124947 028/108] RDMA/cxgb4: Fix incorrect BUG_ON conditions
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Based on original work from Jay Hernandez <jay@chelsio.com>

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 8a9c399eeee8c2d99e22b975f6023001a1fde88f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/cq.c b/drivers/infiniband/hw/cxgb4/cq.c
index 906119f..d6a7db2 100644
--- a/drivers/infiniband/hw/cxgb4/cq.c
+++ b/drivers/infiniband/hw/cxgb4/cq.c
@@ -603,7 +603,7 @@ proc_cqe:
   */
  if (SQ_TYPE(hw_cqe)) {
   int idx = CQE_WRID_SQ_IDX(hw_cqe);
-  BUG_ON(idx > wq->sq.size);
+  BUG_ON(idx >= wq->sq.size);
 
   /*
   * Account for any unsignaled completions completed by
@@ -617,7 +617,7 @@ proc_cqe:
    wq->sq.in_use -= wq->sq.size + idx - wq->sq.cidx;
   else
    wq->sq.in_use -= idx - wq->sq.cidx;
-  BUG_ON(wq->sq.in_use < 0 && wq->sq.in_use < wq->sq.size);
+  BUG_ON(wq->sq.in_use <= 0 && wq->sq.in_use >= wq->sq.size);
 
   wq->sq.cidx = (uint16_t)idx;
   PDBG("%s completing sq idx %u\n", __func__, wq->sq.cidx);
-- 
1.7.1